#Sistemas de ecuaciones lineales

##Métodos Directos

En la figura anterior se muestran cinco reactores interconectados, se necesitan cinco ecuaciones
de balance de masa (flujo x concentración) para representar el sistema.
Sistema de ecuaciones del balance de masa
$$
\begin{array}{ccccccc}
6C_1&+&0&-&C_3&+&0&+&0&=&50\\
3C_1&+&3C_2&+&0&+&0&+&0&=&0\\
0&-&C_2&+&9C_3&+&0&+&0&=&160\\
0&+&C_2&+&8C_3&-&11C_4&+&2C_5&=&0\\
3C_1&+&C_2&+&0&+&0&-&4C_5&=&0\\
\end{array}
$$

In [1]:
import numpy as np
import scipy as sp

In [2]:
#se crea la matriz de coeficientes del sistema a resolver
A=np.array([[6,0,-1,0,0],
            [3,-3,0,0,0],
            [0,-1,9,0,0],
            [0,1,8,-11,2],
            [3,1,0,0,-4]
            ])
#se crea el vector de los terminos independientes
b=np.array([[50,0,160,0,0]])

In [3]:
A

array([[  6,   0,  -1,   0,   0],
       [  3,  -3,   0,   0,   0],
       [  0,  -1,   9,   0,   0],
       [  0,   1,   8, -11,   2],
       [  3,   1,   0,   0,  -4]])

In [4]:
Ab=np.concatenate((A,b.T),axis=1) #se crea la matriz extendida concatenando el vector b como una columna (axis=1)
n=np.size(A,axis=1) #regresa el numero de columnas de A, es decir, el numero de incognitas

In [5]:
Ab

array([[  6,   0,  -1,   0,   0,  50],
       [  3,  -3,   0,   0,   0,   0],
       [  0,  -1,   9,   0,   0, 160],
       [  0,   1,   8, -11,   2,   0],
       [  3,   1,   0,   0,  -4,   0]])

Validar si tiene solucion unica por el teorema de **Rouche-Frobenius**

In [6]:
np.linalg.matrix_rank(Ab)

5

In [7]:
solucion_unica = np.linalg.matrix_rank(A)==np.linalg.matrix_rank(Ab)==n
solucion_unica

True

Aplicando el metodo de **inversa-multiplicacion**

In [8]:
import sympy as sy

In [9]:
#se obtiene la inversa de la matriz de coeficientes A
Ainv=np.linalg.inv(A)
Ainv

array([[ 0.16981132, -0.00628931,  0.01886792,  0.        ,  0.        ],
       [ 0.16981132, -0.33962264,  0.01886792, -0.        , -0.        ],
       [ 0.01886792, -0.03773585,  0.11320755,  0.        ,  0.        ],
       [ 0.06003431, -0.07461407,  0.08747856, -0.09090909, -0.04545455],
       [ 0.16981132, -0.08962264,  0.01886792, -0.        , -0.25      ]])

In [10]:
x=Ainv.dot(b.T) #se multiplica la inversa de A por el vector b
x

array([[11.50943396],
       [11.50943396],
       [19.05660377],
       [16.99828473],
       [11.50943396]])

por lo tanto el conjunto solucion es

* $C_1=11.51$
* $C_2=11.51$
* $C_3=19.06$
* $C_4=17$
* $C_5=11.51$

Balancear la siguiente ecuación química

$$
x_1H_2S + x_2Na_2Cr_2O_7 + x_3H_2SO_4 \rightarrow x_4Cr_2(SO_4)_3 + x_5NaHSO4 + x_6S + x_7 H_2O
$$

Balance por elemento

* $S: x_1+x_3 = 3x_4+x_5+x_6 \Rightarrow x_1+x_3-3x_4-x_5-x_6 = 0$
* $Na: 2x_2 = x_5 \Rightarrow 2x_2 -x_5 = 0$
* $Cr: 2x_2 = 2x_4 \Rightarrow 2x_2–2x_4 = 0$
* $H: 2x_1+2x_3 = x_5+2x_7 \Rightarrow 2x_1+2x_3-x_5-2x_7 = 0$
* $O: 7x_2+4x_3 = 12x_4 +4x_5+x_7 \Rightarrow 7x_2+4x_3-12x_4-4x_5-x_7 = 0$

El sistema de ecuaciones lineales es

$$
\begin{pmatrix}
1& 0& 1& -3& -1& -1& 0\\
0& 2& 0& 0& -1& 0& 0 \\
0& 2& 0& -2& 0& 0& 0 \\
2& 0& 2& 0& -1& 0& -2 \\
0& 7& 4& -12& -4& 0& -1  \\
\end{pmatrix}
$$

In [27]:
import numpy as np
G=np.array([[1 , 0, 1, -3,-1, -1, 0],
            [0 , 2, 0, 0,-1, 0, 0],
            [0 , 2, 0, -2,0, 0, 0],
            [2 , 0, 2, 0,-1, 0, -2],
            [0, 7, 4, -12,-4, 0, -1]])

m=np.array([[0 ,0 ,0 ,0 ,0 ]])

In [5]:
Gm=np.concatenate((G,m.T),axis=1)
n=np.size(G,axis=1)

In [6]:
Gm

array([[  1,   0,   1,  -3,  -1,  -1,   0,   0],
       [  0,   2,   0,   0,  -1,   0,   0,   0],
       [  0,   2,   0,  -2,   0,   0,   0,   0],
       [  2,   0,   2,   0,  -1,   0,  -2,   0],
       [  0,   7,   4, -12,  -4,   0,  -1,   0]])

In [7]:
solucion_unica = np.linalg.matrix_rank(G)==np.linalg.matrix_rank(Gm)==n
solucion_unica

False

In [8]:
solucion_multiple = np.linalg.matrix_rank(G)==np.linalg.matrix_rank(Gm)<n
solucion_multiple

True

In [9]:
import sympy as sy

reduccion,_=sy.Matrix(Gm).rref()
reduccion

Matrix([
[1, 0, 0, 0, 0, -9/16,  -3/16, 0],
[0, 1, 0, 0, 0,   1/4,   -1/4, 0],
[0, 0, 1, 0, 0, 13/16, -17/16, 0],
[0, 0, 0, 1, 0,   1/4,   -1/4, 0],
[0, 0, 0, 0, 1,   1/2,   -1/2, 0]])

Por lo tanto el sistema tiene multiples soluciones y guarda las siguientes proporciones

* $x_1=9/16x_6+3/16x_7$
* $x_2=-1/4x_6+1/4x_7$
* $x_3=-13/16x_6+17/16x_7$
* $x_4=-1/4x_6+1/4x_7$
* $x_5=-1/2x_6+1/2x_7$

el conjunto solucion queda expresado como

$$
(9/16x_6+3/16x_7,-1/4x_6+1/4x_7,-13/16x_6+17/16x_7,-1/4x_6+1/4x_7,-1/2x_6+1/2x_7)
$$

para resolver el problema del balanceo de la reaccion quimica se debe buscar un valor de $x_7$ de tal manera que los valores del resto de las incognitas sea un entero.

por ejemplo $x_7=42$, $x_6=34$




In [46]:
x6=34
x7=42
x1=-x6*reduccion[0,-3]-x7*reduccion[0,-2]
x2=-x6*reduccion[1,-3]-x7*reduccion[1,-2]
x3=-x6*reduccion[2,-3]-x7*reduccion[2,-2]
x4=-x6*reduccion[3,-3]-x7*reduccion[3,-2]
x5=-x6*reduccion[4,-3]-x7*reduccion[4,-2]
print(x1,x2,x3,x4,x5,x6,x7)

27 2 17 2 4 34 42


Por lo tanto la reaccion balanceada es

$$
27H_2S + 2Na_2Cr_2O_7 + 17H_2SO_4 \rightarrow 2Cr_2(SO_4)_3 + 4NaHSO4 + 34S + 42H_2O
$$